node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,348 @@
+	pub fn is_canceled(&self) -> bool {
+		self.inner.is_canceled()
+	}
+
+	/// Verify if the `receiver` is connected to the `sender` [`Self`].
+	pub fn is_connected_to(&self, receiver: &OneshotMeteredReceiver<T>) -> bool {
+		self.inner.is_connected_to(&receiver.inner)
+	}
+}
+
+/// Oneshot receiver, created by [`channel`].
+#[derive(Debug)]
+pub struct OneshotMeteredReceiver<T> {
+	name: &'static str,
+	inner: oneshot::Receiver<(Instant, T)>,
+	/// Soft timeout, on expire a warning is printed.
+	soft_timeout_fut: Option<Fuse<Delay>>,
+	soft_timeout: Duration,
+	/// Hard timeout, terminating the sender.
+	hard_timeout_fut: Option<Delay>,
+	hard_timeout: Duration,
+	/// The first time the receiver was polled.
+	first_poll_timestamp: Option<Instant>,
+	shared_meter: Meter,
+}
+
+impl<T> OneshotMeteredReceiver<T> {
+	pub fn close(&mut self) {
+		self.inner.close()
+	}
+
+	pub fn try_recv(&mut self) -> Result<Option<T>, Error> {
+		match self.inner.try_recv() {
+			Ok(Some((_when, val))) => Ok(Some(val)),
+			Err(e) => Err(Error::from(e)),
+			Ok(None) => Ok(None),
+		}
+	}
+}
+
+impl<T> OneshotMeteredReceiver<T> {
+	pub fn meter(&self) -> Meter {
+		self.shared_meter.clone()
+	}
+
+	fn update_meter(&mut self, sent_at_timestamp: Instant, reason: Reason) {
+		let delta: Duration = sent_at_timestamp.elapsed();
+		let val =
+			if delta.as_millis() > u64::MAX as u128 { u64::MAX } else { delta.as_millis() as u64 };
+		self.shared_meter.first_poll_till_end_in_millis.store(val, Ordering::Release);
+
+		self.shared_meter.reason.store(reason as u8, Ordering::Release);
+	}
+}
+
+impl<T> FusedFuture for OneshotMeteredReceiver<T> {
+	fn is_terminated(&self) -> bool {
+		self.inner.is_terminated()
+	}
+}
+
+impl<T> Future for OneshotMeteredReceiver<T> {
+	type Output = Result<T, Error>;
+
+	fn poll(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Result<T, Error>> {
+		let mut me = self.as_mut();
+
+		if me.first_poll_timestamp.is_none() {
+			me.first_poll_timestamp = Some(Instant::now());
+			me.soft_timeout_fut = Some(Delay::new(me.soft_timeout.clone()).fuse());
+			me.hard_timeout_fut = Some(Delay::new(me.hard_timeout.clone()));
+		}
+
+		let soft_timeout = unsafe {
+			self.as_mut().map_unchecked_mut(|this| this.soft_timeout_fut.as_mut().unwrap())
+		};
+		if soft_timeout.poll(ctx).is_ready() {
+			tracing::warn!("Oneshot `{name}` exceeded the soft threshold", name = &self.name);
+		}
+
+		let hard_timeout = unsafe {
+			self.as_mut().map_unchecked_mut(|this| this.hard_timeout_fut.as_mut().unwrap())
+		};
+		if hard_timeout.poll(ctx).is_ready() {
+			let first_poll_timestamp = self.first_poll_timestamp.unwrap();
+			self.update_meter(first_poll_timestamp, Reason::HardTimeout);
+			return Poll::Ready(Err(Error::HardTimeout(self.hard_timeout.clone())))
+		}
+
+		match Pin::new(&mut self.inner).poll(ctx) {
+			Poll::Pending => Poll::Pending,
+			Poll::Ready(Err(e)) => {
+				let first_poll_timestamp = self.first_poll_timestamp.unwrap();
```

Comment:
4 unwraps detected :fearful:

Reply:
I can add expects, but the flow assures they are initialize from the first call to `poll` onwards.